File: code/rabbitmq/java/client/src/main/java/dev/learn/rabbitmq/L03.java
package dev.learn.rabbitmq;

import static dev.learn.rabbitmq.Broker.*;

import com.rabbitmq.client.AMQP;
import com.rabbitmq.client.BuiltinExchangeType;
import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import java.util.Map;
import java.util.TreeMap;
import java.util.concurrent.CountDownLatch;
import java.util.stream.Collectors;
import tools.jackson.databind.json.JsonMapper;

/** Lesson 3: the order producer and consumer with the Java client, reading and writing the C# program's messages. */
final class L03 {

    private L03() {
    }

    record Order(String id, String sku, int quantity) {
    }

    private static final JsonMapper JSON = JsonMapper.builder().build();

    static void declare(Channel channel) throws Exception {
        channel.exchangeDeclare("orders.topic", BuiltinExchangeType.TOPIC, true);
        channel.queueDeclare("orders.billing", true, false, false, null);
        channel.queueBind("orders.billing", "orders.topic", "order.#");
    }

    static void publish() throws Exception {
        try (Connection connection = connect("l03-java-publisher");
             Channel channel = connection.createChannel()) {
            declare(channel);
            Order[] orders = {new Order("B-1", "TUNER", 1), new Order("B-2", "STRAP", 2)};
            for (Order order : orders) {
                // The properties the C# publisher sets with BasicProperties, through a builder.
                AMQP.BasicProperties properties = new AMQP.BasicProperties.Builder()
                        .contentType("application/json")
                        .messageId(order.id())
                        .type("order.created")
                        .appId("java")
                        .deliveryMode(2)
                        .headers(Map.of("region", "us"))
                        .build();
                byte[] body = JSON.writeValueAsBytes(order);
                channel.basicPublish("orders.topic", "order.created.us", properties, body);
                System.out.println("published " + order.id() + " " + text(body));
            }
        }
    }

    static void consume() throws Exception {
        try (Connection connection = connect("l03-java-consumer");
             Channel channel = connection.createChannel()) {
            declare(channel);
            channel.basicQos(10);
            CountDownLatch waiting = new CountDownLatch((int) channel.messageCount("orders.billing"));

            channel.basicConsume("orders.billing", false,
                    (consumerTag, delivery) -> {
                        AMQP.BasicProperties p = delivery.getProperties();
                        Order order = JSON.readValue(delivery.getBody(), Order.class);
                        String headers = p.getHeaders() == null ? "" : new TreeMap<>(p.getHeaders()).entrySet().stream()
                                .map(h -> h.getKey() + "=" + h.getValue())
                                .collect(Collectors.joining(", "));
                        long tag = delivery.getEnvelope().getDeliveryTag();
                        System.out.println("#" + tag + " " + delivery.getEnvelope().getRoutingKey()
                                + " content-type=" + p.getContentType() + " message-id=" + p.getMessageId()
                                + " type=" + p.getType() + " app-id=" + p.getAppId() + " headers=[" + headers + "]");
                        System.out.println("   " + order);
                        channel.basicAck(tag, false);
                        waiting.countDown();
                    },
                    consumerTag -> { });
            await(waiting, "the waiting orders");
        }
    }
}

File: code/rabbitmq/csharp/L03.cs
using System.Text.Json;
using RabbitMQ.Client;
using RabbitMQ.Client.Events;
using static Learn.Rabbit.Broker;

namespace Learn.Rabbit;

public record Order(string Id, string Sku, int Quantity);

// Lesson 3: a producer and a consumer that agree on a topology and a message format, not on a language.
// The Java client and the Spring AMQP application in ../java read and write the same messages.
static class L03
{
    static readonly JsonSerializerOptions Json = new(JsonSerializerDefaults.Web);

    // Every client declares the same topology with the same settings, so it doesn't matter which one starts first.
    public static async Task DeclareAsync(IChannel channel)
    {
        await channel.ExchangeDeclareAsync("orders.topic", ExchangeType.Topic, durable: true);
        await channel.QueueDeclareAsync("orders.billing", durable: true, exclusive: false, autoDelete: false);
        await channel.QueueBindAsync("orders.billing", "orders.topic", "order.#");
    }

    public static async Task Publish()
    {
        await using IConnection connection = await ConnectAsync("l03-csharp-publisher");
        await using IChannel channel = await connection.CreateChannelAsync();
        await DeclareAsync(channel);

        Order[] orders = [new("A-1", "GUITAR-STRINGS", 2), new("A-2", "CAPO", 1), new("A-3", "PICKS", 12)];
        foreach (var order in orders)
        {
            // BasicProperties is the envelope: metadata the broker and the consumer can read without parsing the body.
            var properties = new BasicProperties
            {
                ContentType = "application/json",
                MessageId = order.Id,
                Type = "order.created",
                AppId = "csharp",
                DeliveryMode = DeliveryModes.Persistent,
                Headers = new Dictionary<string, object?> { ["region"] = "eu" },
            };
            byte[] body = JsonSerializer.SerializeToUtf8Bytes(order, Json);
            await channel.BasicPublishAsync("orders.topic", "order.created.eu", mandatory: false, properties, body);
            Console.WriteLine($"published {order.Id} {Text(body.AsMemory())}");
        }
    }

    public static async Task Consume()
    {
        await using IConnection connection = await ConnectAsync("l03-csharp-consumer");
        await using IChannel channel = await connection.CreateChannelAsync();
        await DeclareAsync(channel);
        // At most 10 unacknowledged messages on this channel at a time (lesson 4).
        await channel.BasicQosAsync(prefetchSize: 0, prefetchCount: 10, global: false);

        var remaining = (int)await channel.MessageCountAsync("orders.billing");
        var done = new TaskCompletionSource();
        if (remaining == 0)
        {
            done.SetResult();
        }

        var consumer = new AsyncEventingBasicConsumer(channel);
        consumer.ReceivedAsync += async (_, delivery) =>
        {
            // The body is only valid during the callback: the client reuses its buffer. Deserialize or copy it here.
            IReadOnlyBasicProperties p = delivery.BasicProperties;
            var order = JsonSerializer.Deserialize<Order>(delivery.Body.Span, Json);
            var headers = p.Headers is null ? "" : string.Join(", ", p.Headers.OrderBy(h => h.Key, StringComparer.Ordinal).Select(h => $"{h.Key}={Text(h.Value)}"));
            Console.WriteLine($"#{delivery.DeliveryTag} {delivery.RoutingKey} content-type={p.ContentType} message-id={p.MessageId} type={p.Type} app-id={p.AppId} headers=[{headers}]");
            Console.WriteLine($"   {order}");

            await channel.BasicAckAsync(delivery.DeliveryTag, multiple: false);
            if (--remaining == 0)
            {
                done.TrySetResult();
            }
        };

        await channel.BasicConsumeAsync("orders.billing", autoAck: false, consumer);
        await WaitAsync(done.Task, "the waiting orders");
    }
}
